frontend/controller/actions/chatroom.js

```diff
@@ -183,6 +183,18 @@ export default (sbp('sbp/selectors/register', {
       }
     }))
   },
+  'gi.actions/chatroom/pubsub/sub': (chatroomContractID: string) => {
+    const pubsub = sbp('okTurtles.data/get', PUBSUB_INSTANCE)
+    pubsub.sub(chatroomContractID, true)
+  },
+  'gi.actions/chatroom/pubsub/unsub': (chatroomContractID: string) => {
+    const pubsub = sbp('okTurtles.data/get', PUBSUB_INSTANCE)
+    pubsub.unsub(chatroomContractID, true)
+  },
```

Comment:
So... we can't do this. The pubsub system is used to listen for events for contracts. Have a look at `'chelonia/connect'`, `'chelonia/pubsub/update'`, and related code in `chelonia.js`.

Only that code is allowed to manage subscribing/unsubscribing from the contractIDs. If any other code starts managing that, it will break the app because the app will miss important new events on the contract chains and will fail to process updates to the group/chatroom state.

Another important point about the code in `'chelonia/connect'`: that code contains some `messageHandlers`, that currently look like this:

```js
    this.pubsub = createClient(pubsubURL, {
      ...this.config.connectionOptions,
      messageHandlers: {
        [NOTIFICATION_TYPE.ENTRY] (msg) {
          // We MUST use 'chelonia/private/in/enqueueHandleEvent' to ensure handleEvent()
          // is called AFTER any currently-running calls to 'chelonia/contract/sync'
          // to prevent gi.db from throwing "bad previousHEAD" errors.
          // Calling via SBP also makes it simple to implement 'test/backend.js'
          const { contractID } = GIMessage.deserializeHEAD(msg.data)
          sbp('chelonia/private/in/enqueueHandleEvent', contractID, msg.data)
        },
        [NOTIFICATION_TYPE.VERSION_INFO] (msg) {
          const ourVersion = process.env.GI_VERSION
          const theirVersion = msg.data.GI_VERSION

          const ourContractsVersion = process.env.CONTRACTS_VERSION
          const theirContractsVersion = msg.data.CONTRACTS_VERSION
          if (ourVersion !== theirVersion || ourContractsVersion !== theirContractsVersion) {
            sbp('okTurtles.events/emit', NOTIFICATION_TYPE.VERSION_INFO, { ...msg.data })
          }
        },
        [REQUEST_TYPE.PUSH_ACTION] (msg) {
          sbp('okTurtles.events/emit', REQUEST_TYPE.PUSH_ACTION, { data: msg.data })
        }
      }
```

We cannot really modify this code to add new handlers here. In fact, these existing handlers need to be modified (and I'll create a separate issue) so that they're not Group Income specific (as Chelonia is a generic framework).

Handlers for different types of messages could be passed in via a parameter to `'chelonia/connect'`, or via `'chelonia/configure'`.

Anyways, just mentioning that in case you need it.

Remember that with @snowteamer's changes you can listen to arbitrary rooms, not just contractIDs. But you might not need to use that I don't think... you could probably just publish a special type of message on the existing pubsub for the contractID, just with a different message type so that it's not classified as an `ENTRY` message and processed.

Reply:
@taoeffect Thanks for the detailed explanation about the system.

> Only that code is allowed to manage subscribing/unsubscribing from the contractIDs.

> Remember that with @snowteamer's changes you can listen to arbitrary rooms, not just contractIDs. 

it sounds like I must avoid using the chatroom contractID as a channel name, and in this case prefixing it something like `CHATROOM_UI-${contractID}` can be an alternative solution. Let me make updates for this.

Reply:
@SebinSong well, I'm not 100% sure. You could use the existing subscriptions that already established for you and publish different types of events on them. That's one simple option. The other option is to manage a new related subscription, which seems more complicated...

Reply:
@taoeffect Right. updated accordingly.